src/counterfactuals/core_struct.jl

```diff
@@ -0,0 +1,155 @@
+"""
+A struct that collects all information relevant to a specific counterfactual explanation for a single individual.
+"""
+mutable struct CounterfactualExplanation <: AbstractCounterfactualExplanation
+    x::AbstractArray
+    target::RawTargetType
+    target_encoded::EncodedTargetType
+    s′::AbstractArray
+    data::DataPreprocessing.CounterfactualData
+    M::Models.AbstractFittedModel
+    generator::Generators.AbstractGenerator
+    generative_model_params::NamedTuple
+    params::Dict
+    search::Union{Dict,Nothing}
+    convergence::Dict
+    num_counterfactuals::Int
+    initialization::Symbol
+end
+
+"""
+	function CounterfactualExplanation(;
+		x::AbstractArray,
+		target::RawTargetType,
+		data::CounterfactualData,
+		M::Models.AbstractFittedModel,
+		generator::Generators.AbstractGenerator,
+		max_iter::Int = 100,
+		num_counterfactuals::Int = 1,
+		initialization::Symbol = :add_perturbation,
+		generative_model_params::NamedTuple = (;),
+		min_success_rate::AbstractFloat=0.99,
+        converge_when::Symbol=:decision_threshold,
+        invalidation_rate::AbstractFloat=0.5,
+        learning_rate::AbstractFloat=1.0,
+        variance::AbstractFloat=0.01,
+	)
+
+Outer method to construct a `CounterfactualExplanation` structure.
+"""
+function CounterfactualExplanation(
+    x::AbstractArray,
+    target::RawTargetType,
+    data::CounterfactualData,
+    M::Models.AbstractFittedModel,
+    generator::Generators.AbstractGenerator;
+    num_counterfactuals::Int=1,
+    initialization::Symbol=:add_perturbation,
+    generative_model_params::NamedTuple=(;),
+    max_iter::Int=100,
+    decision_threshold::AbstractFloat=0.5,
+    gradient_tol::AbstractFloat=parameters[:τ],
+    min_success_rate::AbstractFloat=parameters[:min_success_rate],
+    converge_when::Symbol=:decision_threshold,
+    invalidation_rate::AbstractFloat=0.5,
+    learning_rate::AbstractFloat=1.0,
+    variance::AbstractFloat=0.01,
+)
+
+    # Assertions:
+    @assert any(predict_label(M, data) .== target) "You model `M` never predicts the target value `target` for any of the samples contained in `data`. Are you sure the model is correctly specified?"
```

Comment:
There's a small typo in the error message, it's supposed to be `Your model M (...)` instead of `You model M (...)`

Reply:
I see your point, makes sense to me. Could you please create that issue?

Reply:
I've made an issue for it.